File: compiler-demo/c/pa2s.c
#include <stdint.h>
#include <stdlib.h>

typedef struct {  
	uint8_t r;
	uint8_t g;
	uint8_t b;
	uint8_t a;
} rgba_s;

typedef struct {  
	uint8_t *restrict r;
	uint8_t *restrict g;
	uint8_t *restrict b;
	uint8_t *restrict a;
} rgba_pa;

enum {
	ITEMS = 256
};

rgba_s s[ITEMS] = {0};

uint8_t r[ITEMS] = { [0 ... (ITEMS - 1)] = 'r' };
uint8_t g[ITEMS] = { [0 ... (ITEMS - 1)] = 'g' };
uint8_t b[ITEMS] = { [0 ... (ITEMS - 1)] = 'b' };
uint8_t a[ITEMS] = { [0 ... (ITEMS - 1)] = 'a' };

rgba_pa pa = { r, g, b, a };

static void pa2s(const rgba_pa *restrict pa, rgba_s *restrict s, size_t n)
{
	for (size_t x = 0; x < n; x++) {
		s[x].r = pa->r[x];
		s[x].g = pa->g[x];
		s[x].b = pa->b[x];
		s[x].a = pa->a[x];
	}
}

int main(void)
{
	for (size_t x = 0; x < (1000 * 1000); x++)
		for (size_t y = 0; y < 64; y++)
			pa2s(&pa, s, ITEMS);

	return 0;
}

File: compiler-demo/go/pa2s.go
package main

type (
	rgba_s struct {
		r uint8
		g uint8
		b uint8
		a uint8
	}

	rgba_pa struct {
		r []uint8
		g []uint8
		b []uint8
		a []uint8
	}
)

const (
	ITEMS = 256
	N     = 32 * 1000 * 1000
)

var (
	s [ITEMS]rgba_s

	r, g, b, a [ITEMS]uint8

	pa = rgba_pa{r[:], g[:], b[:], a[:]}
)

func pa2s(pa rgba_pa, s []rgba_s, n int) {
	for i := 0; i < n; i++ {
		s[i].r = pa.r[i]
		s[i].g = pa.g[i]
		s[i].b = pa.b[i]
		s[i].a = pa.a[i]
	}
}

func main() {
	for i := 0; i < N; i++ {
		pa2s(pa, s[:], ITEMS)
	}
}
